File: app/domain/providers/mistral.py
"""
Mistral AI provider implementation using httpx.
"""

from collections.abc import AsyncIterator
from typing import Callable

import httpx

from ...domain.models import LLMResponse
from ...domain.providers.retryable_provider import RetryableProvider
from ...domain.providers.base import Message


class MistralProvider(RetryableProvider):
    """
    Mistral AI provider using httpx directly (no SDK).
    """

    BASE_URL = "https://api.mistral.ai/v1"
    CHAT_COMPLETIONS_ENDPOINT = f"{BASE_URL}/chat/completions"

    def _setup_provider(self) -> None:
        """Setup Mistral-specific attributes."""
        self.name = "mistral"
        self.model = self.config.model
        self.api_key = self.config.api_key
        self._timeout = httpx.Timeout(timeout=60.0, connect=10.0, read=None)

    def _get_retryable_exceptions(self) -> tuple[type[Exception], ...]:
        """Mistral-specific retryable exceptions."""
        return (
            httpx.TimeoutException,
            httpx.ConnectError,
            httpx.NetworkError,
            httpx.ReadError,
        )

    def _execute_chat_sync(self, prompt: str) -> LLMResponse:
        """Execute synchronous chat with Mistral."""
        messages = [
            {"role": "system", "content": "You are a helpful assistant."},
            {"role": "user", "content": prompt},
        ]
        return self._execute_chat_with_messages(messages)

    def _execute_chat_with_messages(
        self,
        messages: list[Message],
        system_prompt: str | None = None,
    ) -> LLMResponse:
        """Execute synchronous chat with messages using Mistral."""
        # Build messages list
        chat_messages: list[dict] = []

        # Add system prompt if provided
        if system_prompt:
            chat_messages.append({"role": "system", "content": system_prompt})

        # Add provided messages
        chat_messages.extend(messages)

        # Request payload
        data = {
            "model": self.model,
            "messages": chat_messages,
            "temperature": self.config.temperature,
            "response_format": {"type": "json_object"},
        }

        headers = {
            "Authorization": f"Bearer {self.api_key}",
            "Content-Type": "application/json",
        }

        with httpx.Client(timeout=self._timeout) as client:
            response = client.post(
                self.CHAT_COMPLETIONS_ENDPOINT,
                headers=headers,
                json=data,
            )
            response.raise_for_status()
            response_data = response.json()

        # Extract response content
        content = response_data["choices"][0]["message"]["content"]

        # Extract usage info
        usage = response_data.get("usage", {})
        prompt_tokens = usage.get("prompt_tokens", 0)
        completion_tokens = usage.get("completion_tokens", 0)

        return self._build_usage_response(
            content=content,
            prompt_tokens=prompt_tokens,
            completion_tokens=completion_tokens,
        )

    def _execute_chat_stream(
        self,
        prompt: str,
        on_chunk: Callable[[str], None] | None = None,
    ) -> AsyncIterator[tuple[str, int, int]]:
        """
        Execute streaming chat with Mistral.

        Yields:
            (content, prompt_tokens, completion_tokens)
            - content is empty string for final yield with token counts
        """
        import json as json_lib

        headers = {
            "Authorization": f"Bearer {self.api_key}",
            "Content-Type": "application/json",
        }
        data = {
            "model": self.model,
            "messages": [
                {"role": "system", "content": "You are a helpful assistant."},
                {"role": "user", "content": prompt},
            ],
            "temperature": self.config.temperature,
            "stream": True,
        }

        prompt_tokens = 0
        completion_tokens = 0

        with httpx.Client(timeout=self._timeout) as client:
            with client.stream(
                "POST",
                self.CHAT_COMPLETIONS_ENDPOINT,
                headers=headers,
                json=data,
            ) as response:
                response.raise_for_status()
                for line in response.iter_lines():
                    if not line or not line.startswith("data: "):
                        continue

                    chunk_data = line[6:]  # Remove "data: " prefix
                    if chunk_data == "[DONE]":
                        break

                    chunk_json = json_lib.loads(chunk_data)
                    delta = chunk_json.get("choices", [{}])[0].get("delta", {}).get(
                        "content", ""
                    )

                    if delta:
                        if on_chunk:
                            on_chunk(delta)
                        yield (delta, 0, 0)

                    usage = chunk_json.get("usage", {})
                    if usage.get("prompt_tokens"):
                        prompt_tokens = usage["prompt_tokens"]
                    if usage.get("completion_tokens"):
                        completion_tokens = usage["completion_tokens"]

        # Yield final response with token counts
        yield ("", prompt_tokens, completion_tokens)

File: ts-agent/src/lib/llm/providers/mistral.ts
// Mistral provider
import { LLMResponse, Message } from "../../../types/llm";
import { LLMInterface } from "../client";
import { configDotenv } from "dotenv";
import { logger } from "../../logger";

configDotenv();

const log = logger.child("llm:mistral");

interface MistralUsage {
  prompt_tokens: number;
  total_tokens: number;
  completion_tokens: number;
}

interface MistralMessage {
  role: string;
  content: string;
}

interface MistralChoice {
  index: number;
  finish_reason: string;
  message: MistralMessage;
}

interface MistralResponse {
  model: string;
  usage: MistralUsage;
  choices: MistralChoice[];
}

export class MistralProvider implements LLMInterface {
  private endpoint: string;

  constructor(
    private baseUrl: string,
    private apiKey: string,
    private model: string,
    private temperature: number,
  ) {
    this.endpoint = `${this.baseUrl}/chat/completions`;
  }

  async chat(messages: Message[], systemPrompt: string): Promise<LLMResponse> {
    messages = [{ role: "system", content: systemPrompt }, ...messages];

    // Request data
    const data = {
      model: this.model,
      messages: messages,
      temperature: this.temperature,

      ...(this.baseUrl.includes("mistral") && {
        response_format: { type: "json_object" },
      }),
    };

    log.debug("api_call_start", {
      model: this.model,
      message_count: data.messages.length,
      temperature: this.temperature,
      json_mode: this.baseUrl.includes("mistral"),
    });

    const startTime = Date.now();

    const response = await fetch(this.endpoint, {
      method: "POST",
      headers: {
        Authorization: `Bearer ${this.apiKey}`,
        "Content-Type": "application/json",
      },
      body: JSON.stringify(data),
    });

    if (!response.ok) {
      const body = await response.text().catch(() => "");
      log.error("api_call_failed", {
        status: response.status,
        body_preview: body.slice(0, 200),
        duration_ms: Date.now() - startTime,
      });
      throw new Error(`LLM client failed: ${response.status}`);
    }

    const raw = (await response.json()) as MistralResponse;

    log.info("api_call_completed", {
      model: raw.model,
      prompt_tokens: raw.usage.prompt_tokens,
      completion_tokens: raw.usage.completion_tokens,
      total_tokens: raw.usage.total_tokens,
      finish_reason: raw.choices[0]?.finish_reason,
      duration_ms: Date.now() - startTime,
    });

    return {
      content: raw.choices[0].message.content,
      usage: {
        prompt_tokens: raw.usage.prompt_tokens,
        completion_tokens: raw.usage.completion_tokens,
        total_tokens: raw.usage.total_tokens,
      },
      cost: { input_cost: 0, output_cost: 0, total_cost: 0 },
      model: raw.model,
      provider: "mistral",
    };
  }
}
